src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/consumer/PersistentServiceProviderConsumerStore.java

```diff
@@ -0,0 +1,317 @@
+import com.thoughtworks.xstream.converters.extended.NamedMapConverter;
+import com.thoughtworks.xstream.io.HierarchicalStreamReader;
+import com.thoughtworks.xstream.io.HierarchicalStreamWriter;
+import hudson.BulkChange;
+import hudson.CopyOnWrite;
+import hudson.XmlFile;
+import hudson.model.Saveable;
+import hudson.model.listeners.SaveableListener;
+import hudson.util.CopyOnWriteMap;
+import hudson.util.XStream2;
+import jenkins.model.Jenkins;
+import jenkins.util.io.OnMaster;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.*;
+import java.util.logging.Logger;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.Optional.ofNullable;
+import static java.util.logging.Level.SEVERE;
+import static java.util.logging.Level.WARNING;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+/**
+ * todo
+ */
+public class PersistentServiceProviderConsumerStore implements ServiceProviderConsumerStore, Saveable, OnMaster {
+
+    @VisibleForTesting
+    static final transient XStream2 CONSUMERS;
+
+    private static final Logger log = Logger.getLogger(PersistentServiceProviderConsumerStore.class.getName());
+
+    static {
+        CONSUMERS = new XStream2();
+        CONSUMERS.registerConverter(new NamedMapConverter(CONSUMERS.getMapper(), "oauth-consumer",
+                "consumer-key", String.class, "consumer-details", Consumer.class), 100);
+    }
+
+    @CopyOnWrite
+    @VisibleForTesting
+    volatile Map<String, Consumer> consumerEntryMap;
+
+    public PersistentServiceProviderConsumerStore() {
+        CONSUMERS.registerConverter(new ConsumerConverter());
+    }
+
+    @Override
+    public void add(Consumer consumer) {
+        requireNonNull(consumer, "consumer");
+        load();
+        if (consumerEntryMap.containsKey(consumer.getKey())) {
+            log.warning(() -> String.format("Consumer with key '%s' already exists.", consumer.getKey()));
+            throw new StoreException("Consumer already exists");
+        }
+        consumerEntryMap.put(consumer.getKey(), consumer);
+        save();
+    }
+
+    @Override
+    public Optional<Consumer> get(String key) {
+        requireNonNull(key, "key");
+        load();
+        return ofNullable(consumerEntryMap.get(key));
+    }
+
+    @Override
+    public void delete(String key) {
+        requireNonNull(key, "key");
+        load();
+        if (consumerEntryMap.remove(key) != null) {
+            save();
+        }
+    }
+
+    public synchronized void load() {
+        if (consumerEntryMap != null) {
```

Comment:
This is the convention used in all Jenkins "stores", including in core and all the plugins I've looked at, and I understood it the same way you mentioned: to avoid unnecessary read from disk on load, since once read, a) the map is never set to null until next Jenkins shutdown, and b) every update calls `save()` synchronized, so the map and XML on disk are generally considered to always be in-sync.
As for loading on construct, I'd be hesitant to do that, e.g. just in case _something_ isn't loaded/initialised yet, either in jenkins core or the plugin and we should wait, and also, I'm not sure what advantage eager loading would have over this synchronized lazy loading.